app/database/item_store.go

```diff
@@ -240,3 +244,193 @@ func (s *ItemStore) isHierarchicalChain(ids []int64) (bool, error) {
 
 	return true, nil
 }
+
+// CheckSubmissionRights checks if the user can submit an answer for the given item (task):
+// 1. If the task is inside a time-limited chapter, the method checks that the task is a part of
+//    the user's active contest (or the user has full access to one of the task's chapters)
+// 2. The method also check that the item (task) exists and is not read-only.
+//
+// Note: This method doesn't check if the user has access to the item.
+// Note 2: This method also can close the user's active contest (or the user's active team contest).
+func (s *ItemStore) CheckSubmissionRights(itemID int64, user *User) (hasAccess bool, reason, err error) {
+	s.mustBeInTransaction() // because it may close a contest
+	recoverPanics(&err)
+
+	hasRights, reason := s.checkSubmissionRightsForTimeLimitedContest(itemID, user)
+	if !hasRights {
+		return hasRights, reason, nil
+	}
+
+	var result []bool
+	mustNotBeError(s.ByID(itemID).Pluck("bReadOnly", &result).Error())
+	if len(result) == 0 {
+		return false, errors.New("no such item"), nil
+	}
+	if result[0] {
+		return false, errors.New("item is read-only"), nil
+	}
+	return true, nil, nil
+}
+
+func (s *ItemStore) checkSubmissionRightsForTimeLimitedContest(itemID int64, user *User) (bool, error) {
+	// TODO: handle case where the item is both in a contest and in a non-contest chapter the user has access to
+
+	// ItemID & FullAccess for time-limited ancestors of the item
+	// for that the user has at least grayed access.
+	// Note that while an answer is always related to a task,
+	// tasks cannot be time-limited, only chapters can.
+	// So, actually here we select time-limited chapters that are ancestors of the task.
+	var contestItems []struct {
+		ItemID     int64 `gorm:"column:idItem"`
+		FullAccess bool  `gorm:"column:fullAccess"`
+	}
+
+	mustNotBeError(s.Visible(user).
+		Select("items.ID AS idItem, fullAccess").
+		Joins("JOIN items_ancestors ON items_ancestors.idItemAncestor = items.ID").
+		Where("items_ancestors.idItemChild = ?", itemID).
+		Where("items.sDuration IS NOT NULL").
+		Group("items.ID").Scan(&contestItems).Error())
+
+	// The item is not time-limited itself and it doesn't have time-limited ancestors the user has access to.
+	// Or maybe the user doesn't have access to the item at all... But we believe in the item token.
+	// Actually, the token's lifetime is two days long, and the access rights might be changed during this time
+	// which is a problem.
+	if len(contestItems) == 0 {
+		return true, nil // The user can submit an answer
+	}
+
+	activeContestErr := errors.New("the contest has not been started yet or has been already finished")
```

Comment:
-> "the contest has not started yet or has already finished"

Reply:
fixed